src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+		uint64_t rebuild_key;
+
+		f_shard = d_list_entry(current, struct failed_shard,
+				       fs_list);
+		l_shard = &layout->ol_shards[f_shard->fs_shard_idx];
+
+		spare_avail = mapless_remap_next_spare(mmap, mop, spare_tgt);
+
+		if (!spare_avail)
+			goto next_fail;
+
+		rebuild_key = (f_shard->fs_shard_idx * 10) + fail_count++;
+		get_rebuild_target(mmap->mmp_map.pl_poolmap, &spare_tgt,
+				crc(key, rebuild_key), dom_used, layout,
+				md);
+
+		/* The selected spare target is down as well */
+		if (pool_target_unavail(spare_tgt)) {
+			D_ASSERTF(spare_tgt->ta_comp.co_fseq !=
+				  f_shard->fs_fseq, "same fseq %u!\n",
+				  f_shard->fs_fseq);
+
+			/* If the spare target fseq > the current object pool
+			* version, the current failure shard will be handled
+			* by the following rebuild.
+			*/
+			if (spare_tgt->ta_comp.co_fseq > md->omd_ver) {
+				D_DEBUG(DB_PL, DF_OID", fseq %d rank %d"
+					" ver %d\n", DP_OID(md->omd_id),
+					spare_tgt->ta_comp.co_fseq,
+					spare_tgt->ta_comp.co_rank,
+					md->omd_ver);
+				spare_avail = false;
+				goto next_fail;
+			}
+
+			/*
+			* The selected spare is down prior to current failed
+			* one, then it can't be a valid spare, let's skip it
+			* and try next spare.
+			*/
+			if (spare_tgt->ta_comp.co_fseq < f_shard->fs_fseq)
+				continue; /* try next spare */
+
+			/*
+			* If both failed target and spare target are down, then
+			* add the spare target to the fail list for remap, and
+			* try next spare..
+			*/
+			if (f_shard->fs_status == PO_COMP_ST_DOWN)
+				D_ASSERTF(spare_tgt->ta_comp.co_status !=
+					  PO_COMP_ST_DOWNOUT,
+					  "down fseq(%u) < downout fseq(%u)\n",
+					  f_shard->fs_fseq,
+					  spare_tgt->ta_comp.co_fseq);
+
+			f_shard->fs_fseq = spare_tgt->ta_comp.co_fseq;
+			f_shard->fs_status = spare_tgt->ta_comp.co_status;
+
+			current = current->next;
+			d_list_del_init(&f_shard->fs_list);
+			remap_add_one(remap_list, f_shard);
+
+			/* Continue with the failed shard has minimal fseq */
+			if (current == remap_list) {
+				current = &f_shard->fs_list;
+			} else {
+				f_tmp = d_list_entry(current,
+						     struct failed_shard,
+						     fs_list);
+				if (f_shard->fs_fseq < f_tmp->fs_fseq)
+					current = &f_shard->fs_list;
+			}
+			continue; /* try next spare */
+		}
+next_fail:
+		fail_count = 0;
+		if (spare_avail) {
+			/* The selected spare target is up and ready */
+			l_shard->po_target = spare_tgt->ta_comp.co_id;
+			l_shard->po_fseq = f_shard->fs_fseq;
+
+			/*
+			* Mark the shard as 'rebuilding' so that read will
+			* skip this shard.
+			*/
+			if (f_shard->fs_status == PO_COMP_ST_DOWN) {
+				l_shard->po_rebuilding = 1;
+				f_shard->fs_tgt_id = spare_tgt->ta_comp.co_id;
+			}
+		} else {
+			l_shard->po_shard = -1;
+			l_shard->po_target = -1;
+		}
+		current = current->next;
+	}
+
+	mapless_remap_dump(remap_list, md, "after remap:");
+	return 0;
+}
+
+static int
+mapless_obj_spec_place_get(struct pl_mapless_map *mmap, daos_obj_id_t oid,
```

Comment:
This function actually only used for getting first shard position? It shares quite a lot common code with ring_obj_spec_place_begin(), please re-organize it a bit to share the code.